### PR TITLE
PostgresParser

### PR DESCRIPTION
**Pull request by haojin2**
This PR is for the new Pg_Parser that use libpg_query for parsing and transform to Peloton parse tree. Important changes includes:
Modified original libpg_query library for direct access to Postgres parse tree
Support for several select queries
Several miscellaneous changes for some bugs in old code

### CODE REVIEW

File: src/include/parser/parsenodes.h [hunk 1]
```diff
@@ -0,0 +1,249 @@
+#include "nodes.h"
```
Comment (apavlo) [inline src/include/parser/parsenodes.h]: You are missing the standard file header

Comment (haojin2) [inline src/include/parser/parsenodes.h]: Fixed

File: src/include/parser/parsenodes.h [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+#include "pg_list.h"
+
+typedef enum SetOperation
+{
```
Comment (apavlo) [inline src/include/parser/parsenodes.h]: This doesn't follow the Peloton style guidelines

Comment (haojin2) [inline src/include/parser/parsenodes.h]: Where can I find that guideline?

File: src/include/parser/pg_parser.h [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+#pragma once
+
+#include "parser/statements.h"
+#include "parser/pg_query.h"
+#include "parser/parser.h"
+#include "parser/parsenodes.h"
+
+namespace peloton {
+namespace pgparser {
```
Comment (haojin2) [inline src/include/parser/pg_parser.h]: changed to parser

File: src/include/parser/nodes.h [hunk 4]
```diff
@@ -0,0 +1,668 @@
...
+ * Portions Copyright (c) 1994, Regents of the University of California
+ *
+ * src/include/parser/nodes.h
+ *
+ *-------------------------------------------------------------------------
+ */
+#ifndef NODES_H
+#define NODES_H
+
+#ifdef __cplusplus
```
Comment (haojin2) [inline src/include/parser/nodes.h]: This was the original code from libpg_query, I can try changing it and see if that causes any problem.

File: test/parser/pg_parser_test.cpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+#include "parser/parser.h"
+
+namespace peloton {
+namespace test {
+
+//===--------------------------------------------------------------------===//
+// Parser Tests
+//===--------------------------------------------------------------------===//
+
+class PgParserTests : public PelotonTest {};
```
Comment (haojin2) [inline test/parser/pg_parser_test.cpp]: Changed this to Pg_ParserTests.

Comment (apavlo) [inline test/parser/pg_parser_test.cpp]: Actually, what if you just renamed it to `PostgresParserTests`? Use `Postgres` instead of `Pg` so that it is more clear.

File: src/include/parser/parsenodes.h (hunk 2 above)
Comment (haojin2) [inline src/include/parser/parsenodes.h]: I could not find any format.py under peloton folder, are there any written guides?

Comment (haojin2) [inline src/include/parser/parsenodes.h]: I guess the format.py has been renamed to formatter.py

Comment (PauloAmora) [inline src/include/parser/parsenodes.h]: There is a formatter at script/formatting/formatter.py

File: third_party/libpg_query/src/pg_query_parse.c [hunk 6]
```diff
@@ -9,6 +9,8 @@
 #include <unistd.h>
 #include <fcntl.h>
 
+#define PARSE_STRING 1
```
Comment (haojin2) [inline third_party/libpg_query/src/pg_query_parse.c]: This is the control flag for enabling debug output for postgres parsetree. The control is now added in the same file on line 78

File: src/include/expression/star_expression.h [hunk 7]
```diff
@@ -38,7 +38,7 @@ class StarExpression : public AbstractExpression {
   virtual void Accept(SqlNodeVisitor *v) { v->Visit(this); }
 
  protected:
-  StarExpression(const FunctionExpression &other) : AbstractExpression(other) {}
+  StarExpression(const AbstractExpression &other) : AbstractExpression(other) {}
```
Comment (jarulraj) [inline src/include/expression/star_expression.h]: Why this change?

File: src/include/parser/postgresparser.h [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+  // transform helper for JoinExpr parsenodes
+  parser::JoinDefinition* JoinTransform(JoinExpr* root);
+
+  // transform helper for from clauses
+  parser::TableRef* FromTransform(List* root);
+
+  // transform helper for select targets
+  std::vector<expression::AbstractExpression*>* TargetTransform(List* root);
+
+  // transform helper for A_Expr nodes
```
Comment (jarulraj) [inline src/include/parser/postgresparser.h]: Good documentation. Might want to expand what A_Expr means etc.

File: src/parser/parser_utils.cpp [hunk 9]
```diff
@@ -110,21 +110,22 @@ void GetExpressionInfo(const expression::AbstractExpression* expr,
     case ExpressionType::STAR:
       inprint("*", num_indent);
       break;
-    case ExpressionType::COLUMN_REF:
-      // TODO: Fix this
-      inprint((expr)->GetExpressionName(), num_indent);
-      //if (expr->GetColumn() != NULL) inprint((expr)->GetColumn(), num_indent);
+    case ExpressionType::VALUE_TUPLE:
+      inprint((expr)->GetInfo().data(), num_indent);
```
Comment (jarulraj) [inline src/parser/parser_utils.cpp]: Can we skip using `inprint` and directly use `LOG_TRACE`? Delete `inprint`.

File: test/parser/postgresparser_test.cpp [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+TEST_F(PostgresParserTests, BasicTest) {
+  std::vector<std::string> queries;
+
+  // Simple select
+  queries.push_back("SELECT * FROM foo;");
+
+  auto parser = parser::PostgresParser::GetInstance();
+  // auto ref_parser = parser::Parser::GetInstance();
+  // Parsing
+  UNUSED_ATTRIBUTE int ii = 0;
```
Comment (jarulraj) [inline test/parser/postgresparser_test.cpp]: Nice job with testing stuff!

File: test/parser/postgresparser_test.cpp [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+  std::vector<std::string> queries;
+
+  // Select with functional call
+  queries.push_back("SELECT COUNT(*) FROM foo;");
+  queries.push_back("SELECT COUNT(DISTINCT id) FROM foo;");
+  queries.push_back("SELECT MAX(*) FROM foo;");
+  queries.push_back("SELECT MIN(*) FROM foo;");
+
+  auto parser = parser::PostgresParser::GetInstance();
+  // auto ref_parser = parser::Parser::GetInstance();
```
Comment (jarulraj) [inline test/parser/postgresparser_test.cpp]: Don't forget to delete these references after removing the old parser.

File: third_party/libpg_query/src/pg_query_parse.c [hunk 12]
```diff
@@ -92,7 +94,11 @@ PgQueryInternalParsetreeAndError pg_query_parse(const char* input)
 
 	parsetree_and_error = pg_query_raw_parse(input);
 
-	pg_query_exit_memory_context(ctx);
+#if PARSE_STRING == 1
```
Comment (jarulraj) [inline third_party/libpg_query/src/pg_query_parse.c]: Adding some comments in this function would help.

File: src/include/expression/star_expression.h (hunk 7 above)
Comment (haojin2) [inline src/include/expression/star_expression.h]: If I remember correctly, this old version reported the following error when I was adding support for star expressions: "error: no matching function for call to ‘peloton::expression::AbstractExpression::AbstractExpression(const peloton::expression::FunctionExpression&)’
"

File: src/include/parser/postgresparser.h (hunk 8 above)
Comment (haojin2) [inline src/include/parser/postgresparser.h]: There're some documentations in the original libpg_query code already. If I add more it's gonna be a bit redundant.

File: test/parser/postgresparser_test.cpp (hunk 11 above)
Comment (haojin2) [inline test/parser/postgresparser_test.cpp]: I'm keeping those for the time being since sometimes I still want to compare with the Peloton parser to check my transformation. I can submit a version without the references but keep those locally, do you prefer that?

File: src/parser/parser_utils.cpp (hunk 9 above)
Comment (haojin2) [inline src/parser/parser_utils.cpp]: I think inprint is a nice helper function since it also handles the number of indents to give a nicer output? We may have to handle the indents manually if we use LOG_TRACE.
